core/src/main/java/org/jboss/jandex/EquivalenceKey.java

```diff
@@ -0,0 +1,571 @@
+package org.jboss.jandex;
+
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Where the common Jandex classes such as {@link ClassInfo}, {@link MethodInfo} or {@link FieldInfo}
+ * provide strict equality, which includes presence or absence of annotations and comparison of their members,
+ * this class provides <em>equivalence</em>. Two Jandex objects are equivalent if and only if they
+ * denote the same declaration or type, without taking into account any annotations. The prime use case
```

Comment:
> Where the common Jandex classes such as {link ClassInfo}, {link MethodInfo} or {link FieldInfo}
 provide strict equality, which includes presence or absence of annotations and comparison of their members,
this class provides <em>equivalence</em>. Two Jandex objects are equivalent if and only if they denote the same declaration or type, without taking into account any annotations.

Hm, I'm not sure I understand these sentences. For example the `ClassInfo` does not define `equals()`/`hashCode()` at all which means the object identity is used. For `FieldInfo` the name, flags, type and annotations are used. And for `MethodInfo` the name, flags, annotations, param names, param types, return type, default value (for annotations) and `receiverType` (I'm not sure what it means ;-). Maybe it wouldn't hurt to list all types of jandex objects and mention the "state" that is taken into account?

Reply:
Fair enough, I'll rephrase this. Forgot about `ClassInfo` not providing equality at all.

Reply:
Rewrote.